tools/single_use/SaveUserIDToInstrumentData.php

```diff
@@ -0,0 +1,147 @@
+    $confirm = true;
+}
+
+echo "\n\nQuerying data...\n\n";
+
+// Get history DB
+$DBInfo = $config->getSetting('database');
+$hDB = isset($DBInfo['historydb']) ? $DBInfo['historydb'] : $DBInfo['database'];
+
+$result_count = 1;
+
+// Query entries in the history table for the latest change
+// made to the flag table, for every CommentID.
+$history = $DB->pselect(
+    "SELECT f.Test_name, f.CommentID, h1.userID
+      FROM flag f
+      JOIN $hDB.history h1 on (f.CommentID=h1.primaryVals)
+      JOIN
+        (
+          SELECT primaryVals, MAX(changeDate) as changeDate
+            FROM $hDB.history
+            WHERE userID <> 'unknown'
+              AND type='U'
+              AND NOT (
+                col='Data_entry' AND old IS NULL
+              )
+            GROUP BY primaryVals
+        ) h2 USING (primaryVals, changeDate)
+            GROUP BY CommentID, h1.userID",
+    array()
+);
+// Loop and index results from history by testname
+$idxHist = [];
+foreach ($history as $entry) {
+    $idxHist[$entry['Test_name']][] = $entry;
+}
+
+foreach(\Utility::getAllInstruments() as $testname => $fullName) {
+    // Instantiate instrument object to get information
+    try {
+        $instrument = \NDB_BVL_Instrument::factory($testname, '', '');
+    } catch (Exception $e) {
+        echo "$testname does not seem to be a valid instrument.\n";
+        continue;
+    }
+
+    // Check if instrument saves data in JSON format i.e. no-SQL table
+    $JSONData = $instrument->usesJSONData();
+    $table = null;
+    if ($JSONData === false) {
+        $table = $instrument->table;
+        if (!$table) {
+            echo "\n\nERROR: The table name for instrument $testname cannot be found.\n";
+            continue;
+        } else if (!$DB->tableExists($table)) {
+            echo "Table $table for instrument $testname does not exist in the Database.\n";
+            continue;
+        } else if (!$DB->columnExists($table, 'UserID')) {
+            echo "Column 'UserID' does not exist in the $table table.\n";
+            continue;
+        }
+    }
+
+    if (empty($idxHist[$testname])) {
+        echo "\n\nThere is no data to import into $testname.\n";
+        continue;
+    } else {
+        echo "\n\nThe following data can be imported into $testname:\n\n";
+        foreach ($idxHist[$testname] as $row) {
+            $commentID = $row['CommentID'];
+            $userID = $row['userID'];
+            echo "\nResult $result_count: $commentID userID: $userID\n";
+            $result_count++;
+        }
+    }
+
+    // Update the instrument table
+    if ($confirm) {
+        echo "\nImporting data into $testname...\n\n";
+
+        // Instantiate instrument at the session level in order to
+        // use _save() function
+        foreach ($idxHist[$testname] as $row) {
+            $commentID = $row['CommentID'];
+            $userID = $row['userID'];
+
+            try {
+                $sessionInst = \NDB_BVL_Instrument::factory(
```

Comment:
This is likely why this script is taking so long. Can we try to use #6869? If that fixes the efficiency, we can either backport the fix to 23 so that it can be used here (slightly complicated/risky since I think it also depends on #6868), or retarget this fix to main for the next release (since it's been broken for so long anyways..)

Reply:
I'm pretty sure 90% of the wait time is the queries on history. I have ran that query separately and its infernal on large projects

Reply:
I did some investigation on my sandbox (which has a much smaller history table), and if that's the case I think we can fix it by temporarily creating an index. I don't know if we would want to keep it or not since the table is frequently written and seldom read, but:

On my sandbox the first time I ran the query it took 45 seconds, the second time it took 40. So in the 40-45s range.

Running: `create index tmptest on history(primaryVals(255))` to add an index took 1.46s. After creating the index, the query took 3.3 seconds. That's down to about 5 seconds total for creating an index and selecting, from 40-45s without an index and assuming a similar proportion for CCNA that would be about 1.5-2h from 14 hours.

(Dropping the index after my tests was pretty much instantaneous..)

Reply:
would we then recommend the index creation in the release notes ?

Reply:
Can we test that my back of the envelope calculation is accurate first? If it is, that seems like a reasonable way to handle it since it's only a problem for large projects..

Reply:
@driusan can you post the command you used to create the index here?

@pierre-p-s can you give it one last try with the index creation and tell us how long it takes for each operation ?

Reply:
it's in my comment above.. `create index tmptest on history(primaryVals(255))`

Reply:
I tried running the CREATE index command @driusan mentionned but after more than 4 hours it failed twice when the SQL server disconnected (ERROR 2013 (`HY000): Lost connection to MySQL server during query`)

Reply:
related issue on redmine: https://redmine.cbrain.mcgill.ca/issues/18993

Reply:
If it's taking 4 hours and then crashing that's more than my back of the envelope calculation had said for index creation, so we should probably just make a note in the release notes that it's slow for large, old projects since it's a 1 time use script.